libavcodec/qsvenc_hevc.c

```diff
@@ -160,6 +161,82 @@ static int generate_fake_vps(QSVEncContext *q, AVCodecContext *avctx)
     return 0;
 }
 
+static int qsv_hevc_set_encode_ctrl(AVCodecContext *avctx,
+                                    const AVFrame *frame, mfxEncodeCtrl *enc_ctrl)
+{
+    if (frame) {
+        AVFrameSideData *sd =
+            av_frame_get_side_data(frame, AV_FRAME_DATA_MASTERING_DISPLAY_METADATA);
+
+        if (sd) {
+            AVMasteringDisplayMetadata *mdm = (AVMasteringDisplayMetadata *)sd->data;
+
+            // SEI is needed when both the primaries and luminance are set
+            if (mdm->has_primaries && mdm->has_luminance) {
+                const int mapping[3] = {1, 2, 0};
+                const int chroma_den = 50000;
+                const int luma_den   = 10000;
+                int i;
+                mfxExtMasteringDisplayColourVolume *mdcv = av_mallocz(sizeof(mfxExtMasteringDisplayColourVolume));
+
+                if (!mdcv)
+                    return AVERROR(ENOMEM);
+
+                mdcv->Header.BufferId = MFX_EXTBUFF_MASTERING_DISPLAY_COLOUR_VOLUME;
+                mdcv->Header.BufferSz = sizeof(*mdcv);
+
+                for (i = 0; i < 3; i++) {
+                    const int j = mapping[i];
+
+                    mdcv->DisplayPrimariesX[i] =
+                        FFMIN(lrint(chroma_den *
+                                    av_q2d(mdm->display_primaries[j][0])),
+                              chroma_den);
+                    mdcv->DisplayPrimariesY[i] =
+                        FFMIN(lrint(chroma_den *
+                                    av_q2d(mdm->display_primaries[j][1])),
+                              chroma_den);
+                }
+
+                mdcv->WhitePointX =
+                    FFMIN(lrint(chroma_den * av_q2d(mdm->white_point[0])),
+                          chroma_den);
+                mdcv->WhitePointY =
+                    FFMIN(lrint(chroma_den * av_q2d(mdm->white_point[1])),
+                          chroma_den);
+
+                mdcv->MaxDisplayMasteringLuminance =
+                    lrint(luma_den * av_q2d(mdm->max_luminance));
```

Comment:
Do we need to clip MaxLuminance to 65535?

Reply:
According to the spec, this value can be greater than 65535.

Reply:
the description in MSDK doc should be wrong.

Reply:
Ok, Got it. Thank you.